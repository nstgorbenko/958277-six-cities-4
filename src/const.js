export const AppRoute = {
  FAVORITES: `/favorites`,
  LOGIN: `/login`,
  MAIN: `/`,
  OFFER: `/offer`,
};

export const CardType = {
  CITIES: `cities`,
  FAVORITES: `favorites`,
  NEAR_PLACES: `near-places`
};

export const ErrorMessage = {
  FAIL_LOAD: [`Oops!`, `Something went wrong.`, `Please refresh page to load data.`],
  NOT_FOUND: [`404`, `Page not found.`]
};

export const OFFER_TYPES = [`apartment`, `room`, `house`, `hotel`];

export const ScreenType = {
  DEFAULT: `default`,
  ERROR: `error`,
  FAVORITES: `favorites`,
  LOGIN: `login`,
  OFFER: `offer`
};

export const SortType = {
  POPULAR: `Popular`,
  PRICE_TO_HIGH: `Price: low to high`,
  PRICE_TO_LOW: `Price: high to low`,
  TOP_RATED: `Top rated first`,
};
