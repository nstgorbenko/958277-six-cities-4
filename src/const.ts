export const MAX_NEARBY_OFFERS_COUNT = 3;
export const MAX_PHOTOS_COUNT = 6;
export const RATING_TITLES: Array<string> = [`perfect`, `good`, `not bad`, `badly`, `terribly`];
export const REVIEWS_TO_SHOW = 10;

export enum AppRoute {
  FAVORITES = `/favorites`,
  LOGIN = `/login`,
  MAIN = `/`,
  OFFER = `/offer`,
}

export enum CardType {
  CITIES = `cities`,
  FAVORITES = `favorites`,
  NEAR_PLACES = `near-places`
}

export const ErrorMessage: {
  [key: string]: Array<string>;
} = {
  FAIL_LOAD: [`Oops!`, `Something went wrong.`, `Please refresh page to load data.`],
  NOT_FOUND: [`404`, `Page not found.`]
};

export const ImageSize: {
  [key: string]: {width: number; height: number};
} = {
  DEFAULT: {
    width: 260,
    height: 200
  },
  FAVORITES: {
    width: 150,
    height: 110
  }
};

export const MarkerSettings: {
  [key: string]: {iconUrl: string; iconSize: [number, number]};
} = {
  DEFAULT: {
    iconUrl: `./img/pin.svg`,
    iconSize: [30, 30]
  },
  ACTIVE: {
    iconUrl: `/img/pin-active.svg`,
    iconSize: [30, 30]
  },
};

export enum PlaceType {
  APARTMENT = `apartment`,
  ROOM = `room`,
  HOUSE = `house`,
  HOTEL = `hotel`,
}

export enum ReviewLength {
  MIN = 50,
  MAX = 300
}

export enum SortType {
  POPULAR = `Popular`,
  PRICE_TO_HIGH = `Price: low to high`,
  PRICE_TO_LOW = `Price: high to low`,
  TOP_RATED = `Top rated first`,
}
