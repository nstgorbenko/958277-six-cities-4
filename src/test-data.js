export const testCity = {
  name: `Amsterdam`,
  location: {
    coordinates: [55.5, 22.2],
    zoom: 10,
  },
};

export const testCities = [{
  name: `Amsterdam`,
  location: {
    coordinates: [55.5, 22.2],
    zoom: 10,
  }
}, {
  name: `Paris`,
  location: {
    coordinates: [77.7, 11.1],
    zoom: 19,
  }
}, {
  name: `Hamburg`,
  location: {
    coordinates: [44.4, 33.3],
    zoom: 11,
  }
}];

export const testReviews = [{
  id: `101`,
  userName: `Alex`,
  userAvatar: `img/avatar-max.jpg`,
  rating: 3.4,
  text: `Just perfect.`,
  time: `2020-03-15T22:55:56.845Z`,
}, {
  id: `102`,
  userName: `Luke`,
  userAvatar: `img/avatar-max.jpg`,
  rating: 5.0,
  text: `Great hospitality, and the breakfast in the morning was a nice touch.`,
  time: `2019-12-05T22:55:56.845Z`,
}];

export const testPlaces = [{
  id: 10,
  city: testCity,
  location: {
    coordinates: [52.3709553943508, 4.90309666406198],
    zoom: 15,
  },
  name: `Stylish apartment in the citycenter`,
  type: `room`,
  description: `Located in the City Center, close to all important attractions.`,
  price: 136,
  photo: `img/apartment-01.jpg`,
  allPhotos: [`img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`],
  bedrooms: 1,
  adults: 2,
  amenities: [`Dishwasher`, `Cabel TV`, `Fridge`],
  host: {
    id: 222,
    name: `Lucy`,
    avatar: `img/avatar-angelina.jpg`,
    isSuper: true,
  },
  rating: 4.5,
  isPremium: false,
  isFavorite: true,
}, {
  id: 11,
  city: testCity,
  location: {
    coordinates: [55.3709553943505, 5.90309666406196],
    zoom: 15,
  },
  name: `Private Attic Studio`,
  type: `apartment`,
  description: `Cosy and comfortable apartment on the ground floor of a typical Amsterdam building.`,
  price: 230,
  photo: `img/apartment-02.jpg`,
  allPhotos: [`img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`],
  bedrooms: 3,
  adults: 4,
  amenities: [`Coffee machine`, `Baby seat`, `Kitchen`],
  host: {
    id: 555,
    name: `Mike`,
    avatar: `img/avatar-max.jpg`,
    isSuper: false
  },
  rating: 4.2,
  isPremium: true,
  isFavorite: false,
}];

export const testServerData = [{
  "id": 10,
  "city": {
    "location": {
      "latitude": 55.5,
      "longitude": 22.2,
      "zoom": 10
    },
    "name": `Amsterdam`
  },
  "location": {
    "latitude": 52.3709553943508,
    "longitude": 4.90309666406198,
    "zoom": 15,
  },
  "title": `Stylish apartment in the citycenter`,
  "type": `room`,
  "description": `Located in the City Center, close to all important attractions.`,
  "price": 136,
  "preview_image": `img/apartment-01.jpg`,
  "images": [`img/apartment-01.jpg`, `img/apartment-01.jpg`, `img/apartment-01.jpg`],
  "bedrooms": 1,
  "max_adults": 2,
  "goods": [`Dishwasher`, `Cabel TV`, `Fridge`],
  "host": {
    "id": 222,
    "name": `Lucy`,
    "avatar_url": `img/avatar-angelina.jpg`,
    "is_pro": true,
  },
  "rating": 4.5,
  "is_premium": false,
  "is_favorite": true,
}, {
  "id": 11,
  "city": {
    "name": `Amsterdam`,
    "location": {
      "latitude": 55.5,
      "longitude": 22.2,
      "zoom": 10,
    },
  },
  "location": {
    "latitude": 55.3709553943505,
    "longitude": 5.90309666406196,
    "zoom": 15,
  },
  "title": `Private Attic Studio`,
  "type": `apartment`,
  "description": `Cosy and comfortable apartment on the ground floor of a typical Amsterdam building.`,
  "price": 230,
  "preview_image": `img/apartment-02.jpg`,
  "images": [`img/apartment-02.jpg`, `img/apartment-02.jpg`, `img/apartment-02.jpg`],
  "bedrooms": 3,
  "max_adults": 4,
  "goods": [`Coffee machine`, `Baby seat`, `Kitchen`],
  "host": {
    "id": 555,
    "name": `Mike`,
    "avatar_url": `img/avatar-max.jpg`,
    "is_pro": false
  },
  "rating": 4.2,
  "is_premium": true,
  "is_favorite": false,
}];

export const testGroupedPlaces = [{
  name: `Amsterdam`,
  location: {
    coordinates: [55.5, 22.2],
    zoom: 10,
  },
  offers: testPlaces
}];
