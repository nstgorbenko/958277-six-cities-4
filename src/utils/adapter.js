export const adaptOffer = (offer) => ({
  id: offer.id,
  city: {
    name: offer.city.name,
    location: {
      coordinates: [offer.city.location.latitude, offer.city.location.longitude],
      zoom: offer.city.location.zoom,
    },
  },
  location: {
    coordinates: [offer.location.latitude, offer.location.longitude],
    zoom: offer.location.zoom,
  },
  name: offer.title,
  type: offer.type,
  description: offer.description,
  price: offer.price,
  photo: offer.preview_image,
  allPhotos: offer.images,
  bedrooms: offer.bedrooms,
  adults: offer.max_adults,
  amenities: offer.goods,
  host: {
    id: offer.host.id,
    name: offer.host.name,
    avatar: offer.host.avatar_url,
    isSuper: offer.host.is_pro,
  },
  rating: offer.rating,
  isPremium: offer.is_premium,
  isFavorite: offer.is_favorite,
});

export const adaptOffers = (offers) =>
  offers.map((offer) => adaptOffer(offer));

export const adaptUserInfo = (info) => ({
  id: info.id,
  name: info.name,
  email: info.email,
  avatar: info.avatar_url,
  isSuper: info.is_pro,
});
