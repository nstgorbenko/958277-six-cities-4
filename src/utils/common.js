export const capitalizeWord = (word) => word[0].toUpperCase() + word.slice(1);

export const getFirstCity = (offers) => {
  const {name, location} = offers[0];
  return ({name, location});
};

export const groupOffersByCities = (offers) => {
  const allCities = offers.reduce((cities, offer) => {
    const currentCity = offer.city.name;
    if (!cities.hasOwnProperty(currentCity)) {
      cities[currentCity] = {
        name: offer.city.name,
        location: offer.city.location,
        offers: [],
      };
    }
    cities[currentCity].offers.push(offer);

    return cities;
  }, {});

  return Object.values(allCities);
};

export const getRatingPercent = (rating) => {
  return Math.round(rating) * 20;
};

export const noop = () => {
  // do nothing
};

export const updateNearbyOffers = (oldNearbyOffers, newNearbyOffer) => {

  if (oldNearbyOffers.length > 0) {
    const newNearbyOfferIndex = oldNearbyOffers.findIndex(({id}) => id === newNearbyOffer.id);
    oldNearbyOffers[newNearbyOfferIndex] = newNearbyOffer;
  }

  return oldNearbyOffers;
};

export const updateOffers = (oldOffers, newOffer) => {
  const city = newOffer.city.name;

  if (oldOffers.length > 0) {
    const cityOffers = oldOffers.filter(({name}) => name === city)[0].offers;
    const newOfferIndex = cityOffers.findIndex(({id}) => id === newOffer.id);
    cityOffers[newOfferIndex] = newOffer;

    return oldOffers;
  }

  return groupOffersByCities([newOffer]);
};

export const updateFavorites = (oldFavorites, newFavorite) => {
  const isFavorite = newFavorite.isFavorite;

  if (isFavorite) {
    oldFavorites.push(newFavorite);
  } else {
    const index = oldFavorites.findIndex(({id}) => id === newFavorite.id);
    oldFavorites.splice(index, 1);
  }

  return oldFavorites;
};
