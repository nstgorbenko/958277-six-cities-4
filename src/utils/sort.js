import {SortType} from "../const.js";

const sortEventsByPriceToHigh = (a, b) => a.price - b.price;

const sortEventsByPriceToLow = (a, b) => b.price - a.price;

const sortEventsByRating = (a, b) => b.rating - a.rating;

export const getSortedPlaces = ([...places], sortType) => {
  switch (sortType) {
    case SortType.POPULAR:
      return places;
    case SortType.PRICE_TO_HIGH:
      return places.sort(sortEventsByPriceToHigh);
    case SortType.PRICE_TO_LOW:
      return places.sort(sortEventsByPriceToLow);
    case SortType.TOP_RATED:
      return places.sort(sortEventsByRating);
    default:
      throw new Error(`Unknown sort type: ${sortType}`);
  }
};
