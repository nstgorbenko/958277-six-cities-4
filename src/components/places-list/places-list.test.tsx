import {BrowserRouter} from "react-router-dom";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import * as React from "react";
import * as renderer from "react-test-renderer";

import {CardType} from "../../const";
import {noop} from "../../utils/common";
import PlacesList from "./places-list";
import {testPlaces} from "../../test-data";

const testStore = configureStore([]);

describe(`PlacesList Component rendering`, () => {
  it(`renders correctly`, () => {
    const store = testStore({
      USER: {authorizationStatus: `AUTH`}
    });

    const tree = renderer
      .create(
          <BrowserRouter>
            <Provider store={store}>
              <PlacesList
                type={CardType.CITIES}
                places={testPlaces}
                sortType="Popular"
                onPlaceCardHover={noop}
              />
            </Provider>
          </BrowserRouter>)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
