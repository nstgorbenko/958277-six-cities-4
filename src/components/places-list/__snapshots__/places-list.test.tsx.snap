// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlacesList Component rendering renders correctly 1`] = `
<div
  className="cities__places-list places__list"
>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
      >
        <img
          alt="Stylish apartment in the citycenter"
          className="place-card__image"
          height={200}
          src="img/apartment-01.jpg"
          width={260}
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            136
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button place-card__bookmark-button--active button"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            In bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "100%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="/offer/10"
          onClick={[Function]}
        >
          Stylish apartment in the citycenter
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Room
      </p>
    </div>
  </article>
  <article
    className="cities__place-card place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="cities__image-wrapper place-card__image-wrapper"
    >
      <a
        href="#"
      >
        <img
          alt="Private Attic Studio"
          className="place-card__image"
          height={200}
          src="img/apartment-02.jpg"
          width={260}
        />
      </a>
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            230
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button  button"
          onClick={[Function]}
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            To bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        <a
          href="/offer/11"
          onClick={[Function]}
        >
          Private Attic Studio
        </a>
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</div>
`;
